### PR TITLE
Node 18.16.0

### PR DESCRIPTION
**Pull request by snickell**
This PR updates Node to the most recent LTS version:
- Node to 18.16.0 and npm to 9.5.1 (the matching version)
- Yarn to a newer patch version 1.22.19

Built off of #49626 but updates to an even more recent node version.

Motivation:
- Node 14 is EOL in 5 days (end of april 2023)
- `RAILS_ENV=test` on non-EOL macOS versions requires maintaining a patch to mini_racer gem version to use libv8-node instead of libv8.

## Deployment checklist

- [x] apps passes tests
- [x] dashboard passes tests
- [x] `docker/dockerfile/Dockerfile` compiles
- [x] drone passes unit tests
- [x] drone passes ui tests
- [x] ad-hoc environment works with node 18 + py3 after chef changes and passes tests
- [x] test starting an ad-hoc from staging and then merging in node-18.16.0 to check converge
- [x] re-merge `yarn.lock` which tends to merge conflict after any package.json changes
- [x] deploy after hours as ad-hoc doesn't guarantee transition
- [x] deployment should NOT occur before May 1 due to AP CS

## Follow-up work

- Potentially worth upgrading selectize 0.12 npm to >0.14 (now renamed to @selectize/selectize), as 0.12 optionally depends on microtime which doesn't compile. Not a big deal, but adds a lot of noise to an otherwise clean `yarn install`, which might hide useful warnings etc in the future. Main reason not to upgrade here is selectize 0.14 pulls in a new jquery-ui version, probably should upgrade main jquery-ui at the same time.
- Ubuntu 18.04 is showing its age (EOL in 1 month I think?). Nodesource binaries are not available for Node > 16 on 18.04, so this PR builds node from source. Worth a quick look-see for how hard it would be to move to Ubuntu `22.04.2` (or even just 20.04, though my instinct is to get ahead of the curve on LTS updates once making a change).

### CODE REVIEW

**Comment by snickell** [comment]
- `npm bin` is removed in node 18, reworking as `npx` commands
- sizeable refactoring of the Dockerfile
   - addresses TODO comments in Dockerfile, e.g. consolidating `apt-get install` commands
   - multi-stage build so node & ruby build in parallel while other deps install
   - changing deps doesn't trigger slloooow rebuild of node & ruby
   - makes it clear what deps are required to compile node and ruby (good for upcoming chef update)
   - uses one consistent style for RUN commands

**Comment by snickell** [comment]
Yay! adhoc / chef changes seem to work:
image

- Nodesource was being used to pull in pre-compiled node binaries, but since Ubuntu 18.04 is almost EOL, Node 18+ is not being built for Ubuntu 18.04. 
- It might make sense to upgrade to Ubuntu 22.04.2 (or 20.04), but that's a rather more sweeping Chef change
- The upstream "node" recipe includes a compile from source mode, but it doesn't pass the required "use GCC 8" flags suggested by Node upstream for compiling on 18.04.
- As a result, I'm building Node from source using a slightly custom recipe
- I've added a conditional to use nodesource by default unless "ubuntu 18.04 + node 18.x" is selected, this should make it resilient to a future programmer upgrading, they won't be stuck on building source, but back on the nodesource path by default.

**Comment by snickell** [comment]
Note: NOT to be merged before May 1, waiting for AP CS to conclude.

Deployment questions for infra:
- Order of operations for drone image tagging: drone images currently still tagged as `:node-18.16.0`, seems in the past habit has been to merge pointing to the named image, and then followup with a PR that switches to claiming the next version number. That make sense here?
- Order of operations for Chef deployment vs merging to source? Because there are other changes, notably to drone, it might make sense to merge source before updating the chef deployment at the right time.
- I'll definitely want assist for the Chef deploy 🙏

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (Hamms) [inline cookbooks/cdo-nodejs/recipes/default.rb]: nit: can this argument be dynamically determined, rather than hardcoded? My understanding is that we generally want this to be equal to the number of processors available to us

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (Hamms) [inline cookbooks/cdo-nodejs/recipes/default.rb:36]: might be worthwhile to do this before installing node from source; if any part of the manual install process overwrites things added by apt, those overrides might be removed by this operation

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (Hamms) [inline cookbooks/cdo-nodejs/recipes/default.rb:21]: tiny nit: any way this value can be determined dynamically rather than hardcoded? My understanding is that it should ideally be equal to the number of processors available to us

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (Hamms) [inline cookbooks/cdo-nodejs/recipes/default.rb:47]: I assume this is where `ark` will put the build artifacts by default? Is there any way this can be  explicitly configured both here and above, rather than relying on defaults which might change?

**Review by Hamms — COMMENTED** [review]
Most of these changes look good, and I dig the organizational improvements to the Dockerfile, but it does make it difficult to parse out in the diff which parts are functional changes to enable the node install and which are just clarity/perf fixes.

I'd love to see the dockerfile fixups extracted out into a separate PR which we can merge in advance of this one, to keep everything isolated!

**Comment by Hamms** [comment]
> Order of operations for drone image tagging: 

Typically what I've done here is wait for the PR to be approved, then upload a new main-version-number-tagged image, make a commit to the branch using that image, and let CI tests pass before merging. We probably could do it as a follow-up, but I like the idea of the staging branch always using a mainline version and only feature branches ever using feature images.

> Order of operations for Chef deployment vs merging to source

Typically the way this works is chef updates are deployed _by_ merging to source. Once this change hits the staging server, it'll build all the cookbooks and persist them to S3 so the later steps in the build pipeline can retrieve and execute them

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (snickell) [reply cookbooks/cdo-nodejs/recipes/default.rb:36]: I think doing the remove after is safer, the ark/source package is installed soley into /usr/local, whereas the apt is in /usr, so they don't step on each-other's toes, and this way, there's always a version of node installed. Not a big deal either way, but was something I mulled a little before going this direction.

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (snickell) [reply cookbooks/cdo-nodejs/recipes/default.rb:47]: great idea, thx!

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (Hamms) [reply cookbooks/cdo-nodejs/recipes/default.rb:36]: Ahh, yeah, that makes sense. In that case, might be worth dropping a in comment that mentions this reasoning!

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (snickell) [reply cookbooks/cdo-nodejs/recipes/default.rb]: I had it coded that way originally, but backed off it, because:
1) The context this will be used in are adhocs, where we want to use ALL the CPU (currently 8, hence the -j 8), and prod/staging where we probably DO NOT want to use all the CPUs, and 8 of 48 felt like a reasonable fraction to use while chef-ing live.
2) The main place where `-j 8` might not be a good value is probably in a local dev kitchen dokker env, but I tested and having -j too high isn't a big negative impact, and if you're gonna have an impact, it feels better to do it on local dev.
3) With that much complexity, the shell got relatively unwieldy and ruined simplicity and clarity for not much gain

**Comment by snickell** [comment]
> I'd love to see the dockerfile fixups extracted out into a separate PR which we can merge in advance of this one, to keep everything isolated!

Feel ya, I was waffling on whether to split these out or not, I'll cherry pick the docker refactor/perf changes into a separate PR.

File: cookbooks/cdo-nodejs/recipes/default.rb
Comment (Hamms) [reply cookbooks/cdo-nodejs/recipes/default.rb]: Hah, yeah, I hadn't considered that what we might actually want here is "the number of processors available but only up to a limit because we sometimes have an unreasonable amount". That makes sense to me! I'd love to see a comment summarizing those reasons

Comment (snickell) [reply cookbooks/cdo-nodejs/recipes/default.rb]: I kinda think this is over-commenting, maybe enough that the PR is linked to this change and has the discuss here? Not gonna fight ya if you got a strong intuition future devs will want to see this 😌

Comment (Hamms) [reply cookbooks/cdo-nodejs/recipes/default.rb]: Nah, no strong intuition. I mention it just because these chef scripts tend to go stale quickly, and we don't have a lot of institutional familiarity with these scripts or with chef in general, so I err way on the side of over-commenting. You're right that it probably is unnecessary but that ain't gonna stop me from at least suggesting it :smile:

**Comment by snickell** [comment]
@Hamms split out the major dockerfile refactor here: https://github.com/code-dot-org/code-dot-org/pull/51630

I'd suggest not deploying that separately but batching the deployment of both of these together.

**Comment by Hamms** [comment]
Is there a specific functional reason you want to batch them together? From a process perspective, it'll be much easier to review the node-version-specific changes in this PR once the other one is merged

**Comment by snickell** [comment]
I went ahead and also updated .drone.yml version in the other PR, so no longer aiming to batch the deploys ✅

(I was assuming we'd merge the other PR first to get a streamlined diff on this one.... my only question was whether to update+deploy `.drone.yml` twice (once for each PR), or just wait until both PRs were merged. Happy to do two deploys ✅✅)

File: docker/dockerfiles/Dockerfile
Comment (snickell) [inline docker/dockerfiles/Dockerfile:62]: blargh, vscode auto-"fixed" my indents on the last PR, hopefully ok to sneak two little indent fixes in here? 🙏

File: docker/setup-compose.yml
Comment (Hamms) [inline docker/setup-compose.yml]: looks like these entries still need to be updated to `codedotorg/code-dot-org:1.7.0`, too

**Review by Hamms — APPROVED** [review]
Looks like there are a couple tiny little snags left to work out, but other than the whitespace fixes, the `yarn.lock` conflicts, and the `*-compose.yml` image tags (and of course assuming this passes all CI tests once those are resolved), this LGTM!

We will probably want to carefully coordinate when and how this gets deployed. The most cautious version of that would involve picking a day to close off staging after the first DTP of the day so we can send this PR through the whole deployment pipeline on its own, with a revert prepped just in case something goes wrong. Given our relatively isolated use of node, however, we might not need to go quite that far and could possibly get away with just making sure the deploy that includes this change is done after our peak traffic for the day has waned.

**Comment by snickell** [comment]
> the `yarn.lock` conflicts

These will probably /also/ have to be re-merged day of the deploy, the yarn+node version changes shifted enough resolutions + output to raise the likelihood of a yarn.lock merge error on any package.json change to "pretty likely". Yarn's builtin merge conflict resolution feature has been doing a great job though, so its typically very easy.

**Comment by davidsbailey** [comment]
Thanks for this upgrade @snickell ! Just wondering in this section about M1 Macs, where there is some mix of mentions of Node 14 and 18. Any idea if this should be Node 18 throughout, or possibly removed entirely (if the problem was indeed specific to Node 14)?

Screenshot 2023-05-11 at 9 59 57 AM

**Comment by cat5inthecradle** [comment]
@davidsbailey let me check on my m1, Node 18 might be supported out of the box.

**Pull request merged**
